### ISSUE TITLE
get_all_triplets_indices returns zero triplets

### ISSUE THREAD

**Issue by marijnl** [KevinMusgrave/pytorch-metric-learning#110]
The `loss_and_miner_utils.get_all_triplets_indices` returns all triplets by default when no miner is provided. If there is no label occurring at least twice then the function returns zero triplets it seems.

I bumping into this when using MPerClassSampler with m=1, my loss kept being 0 because num_non_zero_triplets was constant 0. 
I tested it quickly like this:

```
from pytorch_metric_learning.utils import loss_and_miner_utils
loss_and_miner_utils.get_all_triplets_indices(torch.tensor([1,2,3,4]).to(device)  )
> 0:tensor([], device='cuda:0', dtype=torch.int64)
> 1:tensor([], device='cuda:0', dtype=torch.int64)
> 2:tensor([], device='cuda:0', dtype=torch.int64)

loss_and_miner_utils.get_all_triplets_indices(torch.tensor([1,2,3,4,3]).to(device)  )
> 0:tensor([2, 2, 2, 4, 4, 4], device='cuda:0')
> 1:tensor([4, 4, 4, 2, 2, 2], device='cuda:0')
> 2:tensor([0, 1, 3, 0, 1, 3], device='cuda:0')
```

**Comment by KevinMusgrave** [comment]
This is the intended behavior. But this brings up an important issue:

In TwoStreamMetricLoss, the call to [get_all_triplets_indices](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/trainers/twostream_metric_loss.py#L39) should be ```return lmu.get_all_triplets_indices(labels, labels.clone())```. But why does ```labels``` have to be cloned? It's because of [this comparison](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/utils/loss_and_miner_utils.py#L117). The purpose is to remove the "self" connections if ```ref_labels is labels```. So I think [line 37](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/trainers/twostream_metric_loss.py#L37) also needs to have ```labels.clone()``` passed in as its final argument.

Alternatively, we could add a keyword argument to ```get_all_triplets_indices```. Something like ```same_source```, and then get rid of the ```ref_labels is labels``` comparison.

**Comment by marijnl** [comment]
Check. I misunderstood the function of `get_all_triplet_indices` for what Im trying to do. But thinking of it, it makes perfect sense one cannot make triplets if labels are like `[1,2,3]`

Let my try to implement what you suggest and PR for this

**Comment by marijnl** [comment]
For option 2 this means that for example in TripletMarginMiner when need to explicitly set `get_all_triplets_indices(same_source=False|True)` which is determine by something like `emb is ref_emb` i think. 

Any ideas on doing this a little more efficient? Maybe thats what you refer to as additional logic in`BaseMiner`. Not sure

**Comment by KevinMusgrave** [comment]
Hmm, on second thought, maybe the better solution here is to use ```lmu.get_all_triplets_indices(labels, labels.clone())```, and just add documentation for these functions. This is a pretty specific use-case after all, so it doesn't really make sense to change a bunch of other code to accommodate it.
